### PR TITLE
Fix "per-surface" outputs

### LINKED ISSUE #1336: "Per surface" outputs do not work
The "per surface" problem starts in `COutput::CheckHistoryOutput` where the table header is not given the right number of columns. The function does not add the "per surface names" because we are using the same identifier for normal outputs and "per surface" outputs.
Moreover, the "per surface" implementation is only considering the first marker name and value for printing.
However, there is one value added per "marker analyse".
Let's fix this in a different PR.

_Originally posted by @pcarruscag in https://github.com/su2code/SU2/issues/1226#issuecomment-889067006_

### PR DESCRIPTION
**Pull request by pcarruscag**
Fixes #1336

### CODE REVIEW

File: SU2_CFD/src/output/CFlowOutput.cpp [hunk 1]
```diff
@@ -843,23 +838,24 @@ void CFlowOutput::WriteMetaData(const CConfig *config){
   meta_file.close();
 }
 
-void CFlowOutput::WriteForcesBreakdown(CConfig *config, CGeometry *geometry, CSolver **solver_container){
+void CFlowOutput::WriteForcesBreakdown(const CConfig* config, const CSolver* flow_solver) const {
```
Comment (pcarruscag) [inline SU2_CFD/src/output/CFlowOutput.cpp:841]: Most of that last commit was trying to make this function less ugly.

File: SU2_CFD/src/output/CFlowOutput.cpp [hunk 2]
```diff
@@ -843,23 +838,24 @@ void CFlowOutput::WriteMetaData(const CConfig *config){
...
+  const bool compressible = (config->GetKind_Regime() == ENUM_REGIME::COMPRESSIBLE);
+  const bool incompressible = (config->GetKind_Regime() == ENUM_REGIME::INCOMPRESSIBLE);
+  const bool unsteady = config->GetTime_Domain();
+  const bool viscous = config->GetViscous();
+  const bool dynamic_grid = config->GetDynamic_Grid();
+  const bool gravity = config->GetGravityForce();
+  const bool turbulent = config->GetKind_Solver() == RANS;
+  const bool fixed_cl = config->GetFixed_CL_Mode();
+  const auto Kind_Solver = config->GetKind_Solver();
   const auto Kind_Turb_Model = config->GetKind_Turb_Model();
```
Comment (pcarruscag) [inline SU2_CFD/src/output/CFlowOutput.cpp:856]:
```suggestion
  const auto Kind_Turb_Model = config->GetKind_Turb_Model();
  const bool turbulent = Kind_Turb_Model != NONE;
  const bool fixed_cl = config->GetFixed_CL_Mode();
  const auto Kind_Solver = config->GetKind_Solver();
```

File: SU2_CFD/src/output/COutput.cpp [hunk 3]
```diff
@@ -46,15 +46,17 @@
...
-COutput::COutput(CConfig *config, unsigned short nDim, bool fem_output): femOutput(fem_output) {
+COutput::COutput(const CConfig *config, unsigned short ndim, bool fem_output):
+  rank(SU2_MPI::GetRank()),
+  size(SU2_MPI::GetSize()),
+  nDim(ndim),
+  multiZone(config->GetMultizone_Problem()),
+  gridMovement(config->GetGrid_Movement()),
+  femOutput(fem_output),
+  si_units(config->GetSystemMeasurements() == SI),
+  us_units(config->GetSystemMeasurements() == US) {
```
Comment (TobiKattmann) [inline SU2_CFD/src/output/COutput.cpp:57]: 👍 nice to avoid these repetitions

File: SU2_CFD/src/output/CFlowOutput.cpp [hunk 4]
```diff
@@ -1497,39 +1480,39 @@ void CFlowOutput::WriteForcesBreakdown(const CConfig* config, const CSolver* flo
         break;
 
       case INC_IDEAL_GAS:
-        file << "Fluid Model: INC_IDEAL_GAS " << endl;
-        file << "Variable density incompressible flow using ideal gas law." << endl;
-        file << "Density is a function of temperature (constant thermodynamic pressure)." << endl;
-        file << "Specific heat at constant pressure (Cp): " << config->GetSpecific_Heat_Cp() << " N.m/kg.K." << endl;
-        file << "Molecular weight : " << config->GetMolecular_Weight() << " g/mol" << endl;
-        file << "Specific gas constant: " << config->GetGas_Constant() << " N.m/kg.K." << endl;
+        file << "Fluid Model: INC_IDEAL_GAS \n";
```
Comment (TobiKattmann) [inline SU2_CFD/src/output/CFlowOutput.cpp:1483]: I once saw a video from Mr Turner that using endl to line break is much slower compared to \n. Was that you primary motivation or are there other considerations to this?

Ok it was actually so important to him that it was [Ep 7 of c++ weekly](https://www.youtube.com/watch?v=GMqQOEZYVJQ) back in the good'ol days

**Review by TobiKattmann — APPROVED** [review]
Well thanks for this fix(es) and massive cleanup 💐 Someone at some point will surely appreciate that tedious work!

> Moreover, the "per surface" implementation is only considering the first marker name and value for printing.
> However, there is one value added per "marker analyse".

Can you point me to where that went wrong real quick. I was not able to spot that while going over 😬

File: TestCases/navierstokes/flatplate/lam_flatplate.cfg [hunk 5]
```diff
@@ -87,6 +87,7 @@ MARKER_PLOTTING= ( wall )
 %
 % Marker(s) of the surface where the functional (Cd, Cl, etc.) will be evaluated
 MARKER_MONITORING= ( wall )
+MARKER_ANALYZE= (inlet, outlet)
```
[the remark this answers is not in the document]
Comment (pcarruscag) [reply TestCases/navierstokes/flatplate/lam_flatplate.cfg:90]: We don't seem to.

File: SU2_CFD/include/output/COutput.hpp [hunk 6]
```diff
@@ -57,19 +57,21 @@ class COutput {
...
-  femOutput;            /*!< \brief Boolean to store whether we should use the FEM routines */
+  const bool multiZone;     /*!< \brief Boolean to store whether we are running a multizone problem */
+  const bool gridMovement;  /*!< \brief Boolean to store whether we have grid movement enabled */
+  const bool femOutput;     /*!< \brief Boolean to store whether we should use the FEM routines */
+  const bool si_units;
+  const bool us_units;
 
   /*----------------------------- Screen and history output ----------------------------*/
 
+  const unsigned short fieldWidth = 12; /*!< \brief Width of each column for the screen output (hardcoded for now) */
```
[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/include/output/COutput.hpp:73]: Me neither :) it's the column width it does not affect the history, only screen.

File: SU2_CFD/src/output/CFlowOutput.cpp (hunk 4 above)
Comment (pcarruscag) [reply SU2_CFD/src/output/CFlowOutput.cpp:1483]: Yep that as well, and also clang-format looked nicer.

File: SU2_CFD/src/output/COutput.cpp [hunk 7]
```diff
@@ -1069,44 +1061,48 @@ void COutput::SetScreen_Header(const CConfig *config) {
...
+            PrintingToolbox::PrintScreenScientific(out, field.value, fieldWidth);
+            break;
+          case ScreenOutputFormat::PERCENT:
+            PrintingToolbox::PrintScreenPercent(out, field.value, fieldWidth);
+            break;
+        }
+        (*convergenceTable) << out.str();
       }
     }
-    (*convergenceTable) << out.str();
```
Comment (pcarruscag) [inline SU2_CFD/src/output/COutput.cpp:1109]: We were re-using this string stream and that created the lumped values without separator.

File: SU2_CFD/src/output/COutput.cpp [hunk 8]
```diff
@@ -1069,44 +1061,48 @@ void COutput::SetScreen_Header(const CConfig *config) {
...
+      for (const auto& field : it2->second) {
+        stringstream out;
+        switch (field.screenFormat) {
+          case ScreenOutputFormat::INTEGER:
+            PrintingToolbox::PrintScreenInteger(out, SU2_TYPE::Int(field.value), fieldWidth);
+            break;
+          case ScreenOutputFormat::FIXED:
+            PrintingToolbox::PrintScreenFixed(out, field.value, fieldWidth);
+            break;
+          case ScreenOutputFormat::SCIENTIFIC:
```
Comment (pcarruscag) [inline SU2_CFD/src/output/COutput.cpp:1096]: I should have made this block a lambda or something for re-use, one day...

**Pull request merged**